security/security-spring/src/main/java/org/cbioportal/security/spring/CancerStudyPermissionEvaluator.java

```diff
@@ -0,0 +1,393 @@
+    private static final String MULTIPLE_CANCER_STUDIES_ID = "multiple";
+
+    /**
+     * Implementation of {@code PermissionEvaluator}.
+     *
+     * @param authentication
+     * @param targetId Serialized String cancer study id, 
+     *   String genetic profile id, 
+     *   String sample list id, 
+     *   List<String> of cancer study ids, 
+     *   List<String> of genetic profile ids,
+     *   or List<Integer> of internal sample ids
+     * @param targetType String 'CancerStudy', 
+     *   'SampleList', 
+     *   'GeneticProfile', 
+     *   'SampleList',
+     *   'List<CancerStudyId>', 
+     *   'List<GeneticProfileId>', 
+     *   or 'List<InternalSampleId>'
+     * @param permission
+     */
+    @Override
+    public boolean hasPermission(Authentication authentication, Serializable targetId,
+                                 String targetType, Object permission) {
+        if (log.isDebugEnabled()) {
+            log.debug("hasPermission(), checking permissions on targetId");
+        }
+        if (targetId == null) {
+            return false;
+        }
+        if ("CancerStudy".equals(targetType)) {
+            CancerStudy cancerStudy = studyRepository.getStudy(targetId.toString());
+            if (cancerStudy == null) { 
+                return false;
+            }
+            return hasPermission(authentication, cancerStudy, permission);
+        } else if ("GeneticProfile".equals(targetType)) {
+            GeneticProfile geneticProfile = geneticProfileRepository.getGeneticProfile(targetId.toString());
+            if (geneticProfile == null) {
+                return false;
+            }
+            return hasPermission(authentication, geneticProfile, permission);
+        } else if ("SampleList".equals(targetType)) {
+            SampleList sampleList = sampleListRepository.getSampleList(targetId.toString());
+            if (sampleList == null) {
+                return false;
+            }
+            return hasPermission(authentication, sampleList, permission);
+        }  else if ("List<CancerStudyId>".equals(targetType)) {
+            List<String> studyIds = (List<String>) targetId;
+            for (String studyId : studyIds) {
+                if (!hasPermission(authentication, studyId, "CancerStudy", permission)) {
+                    return false;
+                }
+            }
+            return true;
+        } else if ("List<GeneticProfileId>".equals(targetType)) {
+            List<String> geneticProfileIds = (List<String>) targetId;
+            for (String geneticProfileId : geneticProfileIds) {
+                GeneticProfile geneticProfile = geneticProfileRepository.getGeneticProfile(geneticProfileId);
+                if (geneticProfile == null || !hasPermission(authentication, geneticProfile, permission)) {
+                    return false;
+                }
+            }
+        } else if ("List<InternalSampleId>".equals(targetType)) {
```

Comment:
As I said in the other comment, we don't need this part, and since it gets samples one by one from db for every internal id in `genetic_profile_samples` table (can be thousands of them), it will slow things down.